### PR TITLE
Fixed scanning inverted QRCodes and refactored the way we scan codes

### LINKED ISSUE #3823: Scanning inverted QR code to apply settings does not work
#### Software and hardware versions 
Collect with #3743 

#### Problem description
Scanning inverted QR code to apply settings does not work

#### Steps to reproduce the problem
1. User opens `Configure via QR code`
2. User scans inverted QR code with settings
image

#### Expected behavior
QR code is scanned and settings are applied

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #3823

#### What has been done to verify that this works as intended?
I tested the changes manually.

#### Why is this the best possible solution? Were any other approaches considered?
The problem mentioned in the issue was just a small bug which I fixed. Investigating the code I realized we have a lot of duplicated code in `ScannerWithFlashlightActivity` used by `BarcodeWidget` and `QRScannerFragment` used when we scan settings. I refactored the code and now in both cases we use `QRScannerFragment` (`ScannerWithFlashlightActivity` still exists to hold `QRScannerFragment` but the duplicated code is removed).

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
We need to test scanning settings (no need to check result's just whether the code is scanned or not) and scanning barcodes using the `BarcodeWidget` (with front camera).

#### Do we need any specific form for testing your changes? If so, please attach one.
Any form with `BarcodeWidget` (+ one using front camera ideally).

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [ ] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java [hunk 1]
```diff
@@ -83,6 +83,7 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
 
         Intent intent = new IntentIntegrator(getActivity())
                 .setDesiredBarcodeFormats(getSupportedCodeFormats())
+                .setPrompt(getContext().getString(R.string.barcode_scanner_prompt))
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java]: I think we should make this a separate change for v1.28. Most translators won’t have time to get to this in the next few days so the strings will show up in English where they were translated before.

File: collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java [hunk 2]
```diff
@@ -24,43 +26,70 @@
...
+
+    public enum Action { BARCODE_WIDGET, APPLY_SETTINGS };
+
+    private CaptureManager capture;
+    private DecoratedBarcodeView barcodeScannerView;
     private Button switchFlashlightButton;
     private BeepManager beepManager;
+    private Action action;
+
+    public static QRScannerFragment newInstance(Action action) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java]: Passing in an `Action` to configure the fragment makes it hard to track what the different behavior is between the two types of scanners. Did you consider an abstract `BarcodeScannerFragment` with `SettingsQRCodeScannerFragment` and `BarcodeWidgetScannerFragment` subtypes or something like that? You'd have abstract `handleBarcodeResult` and `getSupportedCodeFormats` methods that are the only things implemented by those concrete classes, making the different behavior really clear.

File: collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java [hunk 3]
```diff
@@ -24,43 +26,70 @@
...
 import org.odk.collect.android.utilities.PermissionUtils;
+import org.odk.collect.android.utilities.WidgetAppearanceUtils;
 
 import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 import java.util.zip.DataFormatException;
 
 public class QRScannerFragment extends Fragment implements DecoratedBarcodeView.TorchListener {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java]: This should be in a different package, now. It's also no longer just a QR code scanner.

**Review by lognaturel — CHANGES_REQUESTED** [review]
Couple of structural suggestions inline.

It'd be great to get tests on the fragment verifying the flashlight and front camera behavior but I think that can follow up.

File: collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java]: Right, fixed.

File: collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java]: We have over 20 translations for that string the default zxing message also is translatable but I noticed it contains less translations so maybe it's better to have this our own?.

File: collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java]: You are right. I Implemented it in that way. Thanks.

File: collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java (hunk 1 above)
Comment (lognaturel) [reply collect_app/src/main/java/org/odk/collect/android/preferences/qr/QRScannerFragment.java]: Oh, right, this string was already in use for the barcode widget. This change just also uses it in the settings QR code context. That makes a lot of sense, thanks.

**Review by lognaturel — APPROVED** [review]
Looks good to me! Great to see duplicated code going away. We can do a separate pass to add automated tests based on this and on the issues found in QA for the QR code scanning improvements.

The refactor looks much better to me. You can see the principle at https://refactoring.com/catalog/replaceConditionalWithPolymorphism.html. I also like the article https://thoughtbot.com/blog/refactoring-replace-conditional-with-polymorphism Clearly there are a lot of violations of this throughout the Collect codebase!

In this case I think it's a simple enough inheritance hierarchy that it feels ok to me. Another option in general to address this kind of issue is to use composition. That's similar to what you were initially going for by passing something into the generic base code scanner. Instead of passing in an enum, though, you'd pass in an object that has some responsibility and can be tested in isolation. The classes would probably look exactly like the subclasses you have now but you'd delegate to the `getSupportedCodeFormats` and `handleScanningResult` methods instead of them being called through polymorphism. We should probably align on which we favor as we get into more of these kinds of refactors.

**Comment by mmarciniak90** [comment]
**Tested with success**

Verified on Android 6.0, 7.0, 8.1, 9.0

Verified cases:
- scan inverted QR code in barcode widget - back and front camera
- scan standard QR code in barcode widget - back and front camera
- scan inverted QR code with settings
- scan standard QR code with settings
- scan invalid, inverted QR code to apply settings

**Comment by kkrawczyk123** [comment]
Verified on Androids 5.1 and 10.

@getodk-bot unlabel "needs testing"
@getodk-bot label "behavior verified"

**Pull request merged**